File: scripts/lib/read_app_version.py
"""Read canonical app version from config/package-solution.json."""

from __future__ import annotations

import json
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
PACKAGE_SOLUTION_PATH = ROOT / "config" / "package-solution.json"
VERSION_PATTERN = re.compile(r"^\d+\.\d+\.\d+\.\d+$")


def read_app_version() -> str:
    data = json.loads(PACKAGE_SOLUTION_PATH.read_text(encoding="utf-8"))
    version = str(data.get("solution", {}).get("version", "")).strip()
    if not VERSION_PATTERN.match(version):
        raise ValueError(f"Invalid solution.version in {PACKAGE_SOLUTION_PATH}: {version or '(empty)'}")
    return version

File: scripts/lib/readAppVersion.mjs
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '../..');
const PACKAGE_SOLUTION_PATH = path.join(ROOT, 'config', 'package-solution.json');

/** Canonical app version from config/package-solution.json (e.g. 1.0.0.0). */
export function readAppVersion() {
  const raw = fs.readFileSync(PACKAGE_SOLUTION_PATH, 'utf8');
  const pkg = JSON.parse(raw);
  const version = String(pkg?.solution?.version || '').trim();
  if (!/^\d+\.\d+\.\d+\.\d+$/.test(version)) {
    throw new Error(`Invalid solution.version in ${PACKAGE_SOLUTION_PATH}: ${version || '(empty)'}`);
  }
  return version;
}

export function readPackageSolution() {
  const raw = fs.readFileSync(PACKAGE_SOLUTION_PATH, 'utf8');
  return JSON.parse(raw);
}

export const REPO_ROOT = ROOT;
